rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -139,43 +152,63 @@ impl FsBuilder {
         self
     }
 
-    /// Returns a [`ByteStream`](crate::byte_stream::ByteStream) from this builder.
-    /// NOTE: If both [`file`](FsBuilder::file) and [`path`](FsBuilder::path) have been called for this FsBuilder, `build` will
-    /// read from the path specified by [`path`](FsBuilder::path).
-    ///
-    /// # Panics
+    /// Specify the offset to start reading from (in bytes)
     ///
-    /// Panics if neither of the `file` or`path` setters were called.
+    /// When used in conjunction with [`length`](FsBuilder::length), allows for reading a single "chunk" of a file.
+    pub fn offset(mut self, offset: u64) -> Self {
+        self.offset = Some(offset);
+        self
+    }
+
+    /// Returns a [`ByteStream`](crate::byte_stream::ByteStream) from this builder.
     pub async fn build(self) -> Result<ByteStream, Error> {
+        if self.path.is_some() && self.file.is_some() {
+            panic!("The 'file' and 'path' options on an FsBuilder are mutually exclusive but both were set. Please set only one")
+        };
+
         let buffer_size = self.buffer_size;
+        let offset = self.offset.unwrap_or(DEFAULT_OFFSET);
+        let length = match self.length {
+            Some(length) => length,
```

Comment:
Should `offset + length` be clamped to the file size if they're greater than it?

For example, if I have an 11 byte file, and a chunk size of 5, the last chunk would be 1 byte. If I passed in 5 as the length for that last chunk out of laziness, should that error, or should that just work?

There's also (maybe?) a use case of uploading a file in chunks that is still being written to by another process, so perhaps it should just let this fly without any manipulation.

Reply:
Yeah, I wasn't sure how strict to be so I decided to just be permissive

Reply:
I came up with a new API based on the poll